crawler.go

```diff
@@ -329,21 +330,33 @@ func (c *Crawler) scanRequestWork(workCh chan chan *http.Request, closeCh chan i
 			spider.reqch <- requestAndChan{req: req, ch: resc}
 			select {
 			case re := <-resc:
-				closeRequest(req)
 				if re.err != nil {
 					c.logf("crawler: send HTTP request got error: %v", re.err)
 				} else {
-					go func(res *http.Response) {
+					go func(clonedReq *http.Request, res *http.Response) {
 						defer closeResponse(res)
 						defer func() {
 							if r := recover(); r != nil {
 								c.logf("crawler: Handler got panic error: %v", r)
 							}
 						}()
+						// TODO(vivek): Don't want to retry forever. How should I escape out?
+						// TODO(vivek): Should I respect Retry-After perfectly?
+						if res.StatusCode == http.StatusTooManyRequests {
+							timeSleep := 10
+							if timeSleepHeader, err := strconv.Atoi(res.Header.Get("Retry-After")); err == nil && timeSleepHeader > 0 && timeSleepHeader < timeSleep {
+								timeSleep = timeSleepHeader
+							}
+							time.Sleep(time.Second * time.Duration(timeSleep))
```

Comment:
You want to respect the context here, so I think this should be something like:

```go
select {
case <- cloneReq.Context().Done():
  c.logf("crawler: aborted because context done")
  return
case <-time.After(timeSleep * time.Second):
  ...
}
```

Reply:
done